File: cpp/test/phonenumbers/geocoding/mapping_file_provider_test.cc
// Copyright (C) 2012 The Libphonenumber Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Author: Patrick Mezard

#include "phonenumbers/geocoding/mapping_file_provider.h"

#include <gtest/gtest.h>  // NOLINT(build/include_order)

#include "phonenumbers/geocoding/geocoding_data.h"

namespace i18n {
namespace phonenumbers {

using std::string;

namespace {

#define COUNTRY_LANGUAGES(code, languagelist)                             \
  const char* country_languages_##code[] = languagelist;                  \
  const CountryLanguages country_##code = {                               \
    country_languages_##code,                                             \
    sizeof(country_languages_##code) / sizeof(*country_languages_##code), \
  };

// Array literals cannot be passed as regular macro arguments, the separating
// commas are interpreted as macro arguments separators. The following dummy
// variadic macro wraps the array commas, and appears as a single argument to an
// outer macro call.
#define ARRAY_WRAPPER(...) __VA_ARGS__

const int country_calling_codes[] = {1, 41, 65, 86};

const int country_calling_codes_size =
  sizeof(country_calling_codes) / sizeof(*country_calling_codes);

COUNTRY_LANGUAGES(1,  ARRAY_WRAPPER({"en"}));
COUNTRY_LANGUAGES(41, ARRAY_WRAPPER({"de", "fr", "it", "rm"}));
COUNTRY_LANGUAGES(65, ARRAY_WRAPPER({"en", "ms", "ta", "zh_Hans"}));
COUNTRY_LANGUAGES(86, ARRAY_WRAPPER({"en", "zh", "zh_Hant"}));

const CountryLanguages* country_languages[] = {
  &country_1,
  &country_41,
  &country_65,
  &country_86,
};

const CountryLanguages* test_get_country_languages(int index) {
  return country_languages[index];
}

}  // namespace

TEST(MappingFileProviderTest, TestGetFileName) {
  MappingFileProvider provider(country_calling_codes,
                               country_calling_codes_size,
                               test_get_country_languages);

  string filename;
  EXPECT_EQ("1_en", provider.GetFileName(1, "en", "", "", &filename));
  EXPECT_EQ("1_en", provider.GetFileName(1, "en", "", "US", &filename));
  EXPECT_EQ("1_en", provider.GetFileName(1, "en", "", "GB", &filename));
  EXPECT_EQ("41_de", provider.GetFileName(41, "de", "", "CH", &filename));
  EXPECT_EQ("", provider.GetFileName(44, "en", "", "GB", &filename));
  EXPECT_EQ("86_zh", provider.GetFileName(86, "zh", "", "", &filename));
  EXPECT_EQ("86_zh", provider.GetFileName(86, "zh", "Hans", "", &filename));
  EXPECT_EQ("86_zh", provider.GetFileName(86, "zh", "", "CN", &filename));
  EXPECT_EQ("", provider.GetFileName(86, "", "", "CN", &filename));
  EXPECT_EQ("86_zh", provider.GetFileName(86, "zh", "Hans", "CN", &filename));
  EXPECT_EQ("86_zh", provider.GetFileName(86, "zh", "Hans", "SG", &filename));
  EXPECT_EQ("86_zh", provider.GetFileName(86, "zh", "", "SG", &filename));
  EXPECT_EQ("86_zh_Hant", provider.GetFileName(86, "zh", "", "TW", &filename));
  EXPECT_EQ("86_zh_Hant", provider.GetFileName(86, "zh", "", "HK", &filename));
  EXPECT_EQ("86_zh_Hant", provider.GetFileName(86, "zh", "Hant", "TW",
                                               &filename));
}

}  // namespace phonenumbers
}  // namespace i18n

File: java/internal/prefixmapper/test/com/google/i18n/phonenumbers/prefixmapper/MappingFileProviderTest.java
/*
 * Copyright (C) 2011 The Libphonenumber Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.i18n.phonenumbers.prefixmapper;

import junit.framework.TestCase;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;
import java.util.SortedMap;
import java.util.TreeMap;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * Unittests for MappingFileProvider.java
 *
 * @author Shaopeng Jia
 */
public class MappingFileProviderTest extends TestCase {
  private final MappingFileProvider mappingProvider = new MappingFileProvider();
  private static final Logger logger = Logger.getLogger(MappingFileProviderTest.class.getName());

  public MappingFileProviderTest() {
    SortedMap<Integer, Set<String>> mapping = new TreeMap<Integer, Set<String>>();
    mapping.put(1, newHashSet("en"));
    mapping.put(86, newHashSet("zh", "en", "zh_Hant"));
    mapping.put(41, newHashSet("de", "fr", "it", "rm"));
    mapping.put(65, newHashSet("en", "zh_Hans", "ms", "ta"));

    mappingProvider.readFileConfigs(mapping);
  }

  private static HashSet<String> newHashSet(String... strings) {
    HashSet<String> set = new HashSet<String>();
    set.addAll(Arrays.asList(strings));
    return set;
  }

  public void testReadWriteExternal() {
    try {
      ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
      ObjectOutputStream objectOutputStream = new ObjectOutputStream(byteArrayOutputStream);
      mappingProvider.writeExternal(objectOutputStream);
      objectOutputStream.flush();

      MappingFileProvider newMappingProvider = new MappingFileProvider();
      newMappingProvider.readExternal(
          new ObjectInputStream(new ByteArrayInputStream(byteArrayOutputStream.toByteArray())));
      assertEquals(mappingProvider.toString(), newMappingProvider.toString());
    } catch (IOException e) {
      logger.log(Level.SEVERE, e.getMessage());
      fail();
    }
  }

  public void testGetFileName() {
    assertEquals("1_en", mappingProvider.getFileName(1, "en", "", ""));
    assertEquals("1_en", mappingProvider.getFileName(1, "en", "", "US"));
    assertEquals("1_en", mappingProvider.getFileName(1, "en", "", "GB"));
    assertEquals("41_de", mappingProvider.getFileName(41, "de", "", "CH"));
    assertEquals("", mappingProvider.getFileName(44, "en", "", "GB"));
    assertEquals("86_zh", mappingProvider.getFileName(86, "zh", "", ""));
    assertEquals("86_zh", mappingProvider.getFileName(86, "zh", "Hans", ""));
    assertEquals("86_zh", mappingProvider.getFileName(86, "zh", "", "CN"));
    assertEquals("", mappingProvider.getFileName(86, "", "", "CN"));
    assertEquals("86_zh", mappingProvider.getFileName(86, "zh", "Hans", "CN"));
    assertEquals("86_zh", mappingProvider.getFileName(86, "zh", "Hans", "SG"));
    assertEquals("86_zh", mappingProvider.getFileName(86, "zh", "", "SG"));
    assertEquals("86_zh_Hant", mappingProvider.getFileName(86, "zh", "", "TW"));
    assertEquals("86_zh_Hant", mappingProvider.getFileName(86, "zh", "", "HK"));
    assertEquals("86_zh_Hant", mappingProvider.getFileName(86, "zh", "Hant", "TW"));
  }
}
